src/libraries/System.Private.CoreLib/src/System/Number.NumberBuffer.cs

```diff
@@ -70,6 +70,28 @@ public void CheckConsistency()
 
                 Debug.Assert(numDigits == DigitsCount, "Null terminator found in unexpected location in Number");
                 Debug.Assert(numDigits < Digits.Length, "Null terminator not found in Number");
+
+                uint totalDigits = (uint)(DigitsCount);
+                uint positiveExponent = (uint)(Math.Max(0, Scale));
+                uint integerDigitsPresent = Math.Min(positiveExponent, totalDigits);
+                uint fractionalDigitsPresent = totalDigits - integerDigitsPresent;
+
+                // Verify that the fractional part does not have any trailing zeros
+                int numberOfTrailingZeros = 0;
+                for (int i = DigitsCount - (int)fractionalDigitsPresent; i < DigitsCount; i++)
+                {
+                    byte digit = Digits[i];
+                    if (digit == 0)
```

Comment:
This should be `digit == '0'` (or take a look at my suggestion above to only do a single check)